### PR TITLE
[flang][cuda] fix parsing of cuda_kernel

### PR DESCRIPTION
**Pull request by ImanHosseini**
Fix parsing of cuda_kernel: it missed a mlir::succeeded check and it was not setting up the `types` and causing mismatch between values and types of the grid/block (CUFKernelValues). @clementval

### CODE REVIEW

File: flang/lib/Optimizer/Dialect/FIROps.cpp [hunk 1]
```diff
@@ -3904,22 +3904,29 @@ mlir::ParseResult parseCUFKernelValues(
     mlir::OpAsmParser &parser,
     llvm::SmallVectorImpl<mlir::OpAsmParser::UnresolvedOperand> &values,
     llvm::SmallVectorImpl<mlir::Type> &types) {
-  if (mlir::succeeded(parser.parseOptionalStar()))
+  if (mlir::succeeded(parser.parseOptionalStar())) {
```
Comment (clementval) [inline flang/lib/Optimizer/Dialect/FIROps.cpp:3907]: No braces on simple if. Lowering and dialect uses MLIR style.

File: flang/lib/Optimizer/Dialect/FIROps.cpp [hunk 2]
```diff
@@ -3904,22 +3904,29 @@ mlir::ParseResult parseCUFKernelValues(
...
+    for (size_t i = 0; i < values.size(); i++) {
+      types.emplace_back(builder.getType<mlir::IntegerType>(32));
+    }
     if (parser.parseRParen())
       return mlir::failure();
   } else {
     if (parser.parseOperand(values.emplace_back()))
       return mlir::failure();
+    auto builder = parser.getBuilder();
+    types.emplace_back(builder.getType<mlir::IntegerType>(32));
```
Comment (clementval) [inline flang/lib/Optimizer/Dialect/FIROps.cpp:3929]:
```suggestion
    types.emplace_back(builder.getI32Type());
```

File: flang/test/Lower/CUDA/cuda-kernel-loop-directive.cuf [hunk 3]
```diff
@@ -1,4 +1,5 @@
 ! RUN: bbc -emit-hlfir -fcuda %s -o - | FileCheck %s
+! RUN: bbc -emit-hlfir -fcuda %s -o - | fir-opt | FileCheck %s
```
Comment (clementval) [inline flang/test/Lower/CUDA/cuda-kernel-loop-directive.cuf:2]: Thanks for adding this here. It'll make the test more robust.

**Review by clementval — APPROVED** [review]
Just couple of nits for the style. Otherwise LGTM.

File: flang/test/Lower/CUDA/cuda-kernel-loop-directive.cuf (hunk 3 above)
Comment (ImanHosseini) [reply flang/test/Lower/CUDA/cuda-kernel-loop-directive.cuf:2]: learned form your previous PR.

**Pull request merged**